Lushu/src/main/kotlin/lushu/ContextGrammar/Grammar/Rules.kt

```diff
@@ -0,0 +1,250 @@
+                mutableTokens[index] = asteriskSymbol.repeat(mutableTokens[index].length)
+            }
+
+            return matchTokensAgainstPatternContext(inputTokens, mutableTokens, index + 1, nextNode)
+        }
+    }
+
+    /**
+     * Finds the matching indices of tokens in the input list based on a list of regular expressions.
+     * @param inputTokens The list of tokens to search for matches.
+     * @param regexList The list of regular expressions to match against the tokens.
+     * @return A list of unique and sorted indices of the matching tokens.
+     * Example:
+     * Input: ["error", "warning", "exception", "success", "invalid"], ["e.", "s.", "c."]
+     * Output: [0, 2, 3]
+     */
+    private fun findMatchingIndex(inputTokens: List<String>, regexList: List<String>): List<Int> {
+        val matchingIndex = regexList.flatMap { regex ->
+            val pattern = Pattern.compile(regex)
+            inputTokens.mapIndexedNotNull { index, word ->
+                val matcher = pattern.matcher(word)
+                if (matcher.find()) index else null
+            }
+        }
+        return matchingIndex.distinct().sorted()
+    }
+
+    /**
+     * Converts a list of tokens into cipher tokens.
+     * This function iterates over the input list and checks if it contains any text patterns that need to be encrypted.
+     * The resulting cipher tokens are returned as a list.
+     * @param inputTokens The list of input tokens to be converted into cipher tokens.
+     * @return The list of cipher tokens generated from the input tokens.
+     * Example:
+     * Let a context "host ip: 123.123.123.123" where 123.123.123.123 needs to be encrypted.
+     * Input: ["server", "host", "ip:",  "123.123.123.123", "is", "online"]
+     * Output: ["server", "host", "ip:",  "***************", "is", "online"]
+     */
+    fun tokens2CipherTokens(inputTokens: List<String>): List<String> {
+        val regexList: List<String> = root.getChildren().map { child -> child.getRegex() }
+        val matchingIndex: List<Int> = findMatchingIndex(inputTokens, regexList)
+
+        var i = defaultMinimalIndex
+
+        val cipherTokens = inputTokens.toMutableList()
+
+        matchingIndex.forEach { index ->
+            if (i <= index) {
+                i = matchTokensAgainstPatternContext(cipherTokens, inputTokens.toMutableList(), index, root)
+            }
+        }
+
+        return cipherTokens
+    }
+
+    /**
+     * Adds rules based on the given context rule by recursively processing each word in the context rule.
+     * @param contextRule The mutable list of strings representing the context rule.
+     * @param current The current node in the tree structure. Defaults to the root node.
+     *
+     * - Example:
+     * Input: mutableListOf("This", "<t>is", "an</t>", "<s>example</s>", "sentence.")
+     */
+    private fun addContextToGrammar(contextRule: MutableList<String>, current: Node? = root) {
```

Comment:
Why is this function called `addContextToGrammar`? Maybe we should rename it to something like `addContextRule`.